### PR TITLE
🚩 feat(Kotlin-DSL) Use the kotlin DSL to provide the instance of Cache

### PR DESCRIPTION
**Pull request by lhenriquegomescamilo**
### Description
The main goal of this PR is to provide a `Cache` instance using the Kotlin DSL, take a look at the following example:
* Before
```koltin
val cache = Cache.Builder()
      .fakeTimeSource(fakeTimeSource)
      .expireAfterWrite(1.minutes)
      .build<Long, String>()
```
* After
```kotlin
    val cache = buildCache<Long, String> { 
        fakeTimeSource(fakeTimeSource)
        expireAfterWrite(1.minutes)
  }
```

* But, if the developers want to use the default configuration, they can use:
```kotlin
val cache =  buildDefaultCache<Long, String>()
```

## Motivations
* https://kotlinlang.org/docs/type-safe-builders.html
* https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/build-list.html

### CODE REVIEW

File: cache4k/src/commonMain/kotlin/io/github/reactivecircus/cache4k/CacheConfig.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package io.github.reactivecircus.cache4k
+
+public fun <K : Any, V : Any> buildCache(init: Cache.Builder.() -> Unit): Cache<K, V> {
+    val cacheBuilder = Cache.Builder()
+    cacheBuilder.init()
+    return cacheBuilder.build()
+}
+
+public fun <K : Any, V : Any> defaultBuildCache(): Cache<K, V> = buildCache {}
```
Comment (hoc081098) [inline cache4k/src/commonMain/kotlin/io/github/reactivecircus/cache4k/CacheConfig.kt:9]: IMO, it's should be `buildDefaultCache`

File: README.md [hunk 2]
```diff
@@ -130,9 +135,7 @@ To set the maximum time an entry can live in the cache since the last write (als
 time-to-live**), where "write" means **adding a new cache entry** or **replacing an existing entry with a new one**:
 
 ```kotlin
-val cache = Cache.Builder()
-    .expireAfterWrite(30.minutes)
-    .build<Long, String>()
+val cache = cacheCOnfig<Long, String> { expireAfterWrite(30.minutes) }
```
Comment (hoc081098) [inline README.md:138]: `buildCache`

File: cache4k/src/commonMain/kotlin/io/github/reactivecircus/cache4k/CacheConfig.kt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package io.github.reactivecircus.cache4k
+
+public fun <K : Any, V : Any> buildCache(init: Cache.Builder.() -> Unit): Cache<K, V> {
```
Comment (hoc081098) [inline cache4k/src/commonMain/kotlin/io/github/reactivecircus/cache4k/CacheConfig.kt:3]: What's about `inline fun`?

**Comment by ychescale9** [comment]
Thanks for the PR!

I generally prefer simple builder APIs unless I need to express the structure and ordering of elements.

https://colinwhite.medium.com/prefer-builders-over-dsls-3dc7fc375d47
https://github.com/apollographql/apollo-kotlin/issues/3301

I feel like DSL is a bit overkill for our usecase and implementing it on top of existing builder APIs increases the API surface of the library without bringing significant improvements to developer experience.

Happy to hear what others think 😃

File: cache4k/src/commonMain/kotlin/io/github/reactivecircus/cache4k/CacheConfig.kt (hunk 1 above)
Comment (lhenriquegomescamilo) [reply cache4k/src/commonMain/kotlin/io/github/reactivecircus/cache4k/CacheConfig.kt]: You're right, that is a better name.

File: README.md (hunk 2 above)
Comment (lhenriquegomescamilo) [reply README.md]: OMG... I lost this typo, thank you!

File: cache4k/src/commonMain/kotlin/io/github/reactivecircus/cache4k/CacheConfig.kt (hunk 3 above)
Comment (lhenriquegomescamilo) [reply cache4k/src/commonMain/kotlin/io/github/reactivecircus/cache4k/CacheConfig.kt:3]: That's right!

**Comment by lhenriquegomescamilo** [comment]
> Thanks for the PR!
>
> I generally prefer simple builder APIs unless I need to express the structure and ordering of elements.
>
> https://colinwhite.medium.com/prefer-builders-over-dsls-3dc7fc375d47 [apollographql/apollo-kotlin#3301](https://github.com/apollographql/apollo-kotlin/issues/3301)
>
> I feel like DSL is a bit overkill for our usecase and implementing it on top of existing builder APIs increases the API surface of the library without bringing significant improvements to developer experience.
>
> Happy to hear what others think smiley


First, happy new year 🥳 

Well, using the Kotlin DSL it's bringing to the library a concept of Kotlin way to solve the problems, for example, the future of the Spring Framework(backend written in Kotlin)  is to use the Kotlin DSL to set up your beans in the context, for example:
* https://spring.io/blog/2020/05/28/spring-fu-0-3-0-and-beyond 
```kotlin
val app = webApplication {
	beans {
		bean<SampleService>()
		bean<SampleHandler>()
	}
	webMvc {
		port = if (profiles.contains("test")) 8181 else 8080
		router {
			val handler = ref<SampleHandler>()
			GET("/", handler::hello)
			GET("/api", handler::json)
		}
		converters {
			string()
			jackson {
				indentOutput = false
			}
		}
	}
}

fun main() {
	app.run()
}
```
You can see these principles as well in the Ktor:
* https://ktor.io/
```kotlin
fun main() {
	embeddedServer(Netty, port = 8000) {
		routing {
			get ("/") {
				call.respondText("Hello, world!")
			}
		}
	}.start(wait = true)
}
```
And if you are working with android, maybe you are using the android jetpack compose:
* https://developer.android.com/jetpack/compose
```kotlin
@Composable
fun JetpackCompose(){
   Card {
      val expanded by remember { multableStateOf(false) }
      Column(Modifier.clickable { expanded = !expanded }) {
          Image(painterResource(R.drawable.jetback_compose))
          AnimatedVisibility(expanded) {
             Text(
                 text = "Jetpack Compose",
                 style = MaterialTheme.typography.h2,
              )
          }
      }
    }
} 
```
And, I really agree with you, but my point for this PR is to provide a builder by DSL following the guideline(type-safe builders) of Kotlin, trying to be more expressive for the other developers.

What do you think guys?

**Comment by ychescale9** [comment]
Closing this for now, if anyone prefers DSL syntax please don't hesitate to open an issue or discussion.

**Pull request closed without merging**